opal/mca/common/ucx/common_ucx_wpool.c

```diff
@@ -949,136 +687,83 @@ static int _tlocal_ctx_connect(_tlocal_ctx_t *ctx_rec, int target)
-        if (mem_rec->mem->rkeys[i]) {
-            ucp_rkey_destroy(mem_rec->mem->rkeys[i]);
+        if (mem_rec->rkeys[i]) {
+            ucp_rkey_destroy(mem_rec->rkeys[i]);
         }
     }
-    free(mem_rec->mem->rkeys);
-
-    /* Remove myself from the memory context structure
-     * This may result in context release as we are using
-     * delayed cleanup */
-    _common_ucx_mem_signout(mem_rec->gmem);
-
-    /* Release fast-path pointers */
-    if (NULL != mem_rec->mem_tls_ptr) {
-        free(mem_rec->mem_tls_ptr);
-    }
+    free(mem_rec->rkeys);
 
-    assert(mem_rec->ctx_rec != NULL);
-    OPAL_ATOMIC_ADD_FETCH32(&mem_rec->ctx_rec->refcnt, -1);
-    assert(mem_rec->ctx_rec->refcnt >= 0);
+    /* Remove item from the list */
+    opal_mutex_lock(&mem_rec->gmem->mutex);
+    opal_list_remove_item(&mem_rec->gmem->mem_records, &mem_rec->super);
+    opal_mutex_unlock(&mem_rec->gmem->mutex);
 
-    free(mem_rec->mem);
+    OBJ_RELEASE(mem_rec);
 
-    memset(mem_rec, 0, sizeof(*mem_rec));
+    return;
 }
 
-static _tlocal_mem_t *_tlocal_add_mem(_tlocal_table_t *tls,
-                                       opal_common_ucx_wpmem_t *mem)
+static _mem_record_t *_tlocal_add_mem_rec(opal_common_ucx_wpmem_t *mem, _ctx_record_t *ctx_rec)
 {
-    size_t i, free_idx = -1;
-    _tlocal_ctx_t *ctx_rec = NULL;
-    int rc = OPAL_SUCCESS, found = 0;
-
-    /* Try to find available spot in the table */
-    for (i=0; i<tls->mem_tbl_size; i++) {
-        if (NULL != tls->mem_tbl[i]->gmem) {
-            if (tls->mem_tbl[i]->gmem->released) {
-                /* Found a dirty record. Need to clean it first */
-                _tlocal_mem_record_cleanup(tls->mem_tbl[i]);
-            }
-        }
-        if ((NULL == tls->mem_tbl[i]->gmem) && !found) {
-            /* Found a clear record */
-            free_idx = i;
-            found = 1;
-        }
-    }
-
-    if (!found){
-        free_idx = tls->mem_tbl_size;
-        rc = _tlocal_tls_memtbl_extend(tls, 4);
-        if (rc != OPAL_SUCCESS) {
-            //TODO: error out
-            return NULL;
-        }
-    }
-
-    tls->mem_tbl[free_idx]->gmem = mem;
-    tls->mem_tbl[free_idx]->mem = calloc(1, sizeof(*tls->mem_tbl[free_idx]->mem));
-
-    ctx_rec = _tlocal_ctx_search(tls, mem->ctx);
-    if (NULL == ctx_rec) {
-        // TODO: act accordingly - cleanup
+    int rc = OPAL_SUCCESS;
+    _mem_record_t *mem_rec = OBJ_NEW(_mem_record_t);
+    if (NULL == mem_rec) {
         return NULL;
     }
 
-    tls->mem_tbl[free_idx]->ctx_rec = ctx_rec;
-    OPAL_ATOMIC_ADD_FETCH32(&ctx_rec->refcnt, 1);
-
-    tls->mem_tbl[free_idx]->mem->worker = ctx_rec->winfo;
-    tls->mem_tbl[free_idx]->mem->rkeys = calloc(mem->ctx->comm_size,
-                                         sizeof(*tls->mem_tbl[free_idx]->mem->rkeys));
+    opal_mutex_lock(&mem->mutex);
+    opal_list_append(&mem->mem_records, &mem_rec->super);
+    opal_mutex_unlock(&mem->mutex);
```

Comment:
Must go after initialization. Or during the flush someone might try to access it before it is initialized
```suggestion
```